web/blog/parquet-and-feather-writing-security-telemetry/index.qmd

```diff
@@ -0,0 +1,868 @@
+
+How many events do we have per schema?
+
+```{r}
+#| code-fold: true
+#| label: number-of-events-by-schema
+schemas <- normalized |>
+  # Pick one element from the run matrix.
+  filter(store == "feather" & zstd.level == 1) |>
+  group_by(schema) |>
+  summarize(n = sum(num_events),
+            bytes_memory = sum(bytes_memory))
+
+schemas |>
+  separate(schema, c("module", "type"), remove = FALSE) |>
+  ggplot(aes(x = reorder(schema, -n), y = n, fill = module)) +
+    geom_bar(stat = "identity") +
+    scale_y_log10(labels = scales::label_comma()) +
+    labs(x = "Schema", y = "Number of Events", fill = "Module") +
+    theme(axis.text.x = element_text(angle = -90, size = 8, vjust = 0.5, hjust = 0))
+```
+
+The above plot (log-scaled y-axis) shows how many events we have per type.
+Between 1 and 100M events, we almost see everything.
+
+What's the typical event size?
+
+```{r}
+#| code-fold: true
+#| label: event-size-by-schema
+schemas |>
+  separate(schema, c("module", "type"), remove = FALSE) |>
+  ggplot(aes(x = reorder(schema, -n), y = bytes_memory / n, fill = module)) +
+    geom_bar(stat = "identity") +
+    guides(fill = "none") +
+    scale_y_continuous(labels = scales::label_bytes(units = "auto_si")) +
+    labs(x = "Schema", y = "Bytes (in-memory)", color = "Module") +
+    theme(axis.text.x = element_text(angle = -90, size = 8, vjust = 0.5, hjust = 0))
+```
+
+The above plot keeps the x-axis from the previous plot, but exchanges the y-axis
+to show normalized event size, in memory after parsing. Most events
+take up a few 100 bytes, with packet data consuming a bit more, and one 5x
+outlier: `suricata.ftp`.
+
+Such distributions are normal, even with these outliers. Some telemetry events
+simply have more string data that's a function of user input. For `suricata.ftp`
+specifically, it can grow linearly with the data transmitted. Here's a stripped
+down example of an event that is greater than 5 kB in its raw JSON:
+
+```json
+{
+  "timestamp": "2021-11-19T05:08:50.885981+0100",
+  "flow_id": 1339403323589433,
+  "pcap_cnt": 5428538,
+  "event_type": "ftp",
+  "src_ip": "10.5.5.101",
+  "src_port": 50479,
+  "dest_ip": "62.24.128.228",
+  "dest_port": 110,
+  "proto": "TCP",
+  "tx_id": 12,
+  "community_id": "1:kUFeGEpYT1JO1VCwF8wZWUWn0J0=",
+  "ftp": {
+    "completion_code": [
+      "155",
+      ...
+      <stripped 330 lines>
+      ...
+      "188",
+      "188",
+      "188"
+    ],
+    "reply": [
+      " 41609",
+      ...
+      <stripped 330 lines>
+      ...
+      " 125448",
+      " 126158",
+      " 29639"
+    ],
+    "reply_received": "yes"
+  }
+}
+```
+
+This matches our mental model. A few hundred bytes per event with some outliers.
+
+### Batching
+
+On the inside, a store is just a concatenation of Arrow record batches. In
+Feather they are literally concatenated sequentially, whereas Parquet
+concatenates them column-wise. Most stores only consist of a few record
+batches. PCAP is the only difference.
```

Comment:
This is not true on either front.

Feather is essentially the wire format of record batches, but that is not a literal concatenation. Instead, schemas and dictionaries are only included when they change, so for our stores just once in the beginning. In order to access a given row in a Feather file, you always need to start at the beginning, iterating over batch by batch until you arrive at the desired batch, which you then materialize before you can access the desired row in the batch via random access.

Parquet has a notion of row groups, which are much like a record batch, except that they are created on write, so their size is set by Parquet rather than the incoming data. Parquet offers random access over both the row groups and within an individual batch that is materialized from a row group. The on-disk layout of Parquet is still row-group by row-group, and in that column by column, so there's no big difference between Parquet and Feather in that regard. Parquet encodes columns using different encoding techniques than Arrow's IPC format.

Reply:
Thanks for the detailed explanation. I added that into the section in [4939950](https://github.com/tenzir/vast/pull/2617/commits/49399507cc73a07a82ab793c4a5255e79214f7e5).

Reply:
If this reads well, feel free to resolve.

Reply:
Yes, this is great!